src/coreclr/System.Private.CoreLib/src/Internal/Runtime/InteropServices/InMemoryAssemblyLoader.cs

```diff
@@ -2,48 +2,85 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 
 using System;
+using System.Diagnostics.CodeAnalysis;
 using System.Runtime.InteropServices;
 using System.Runtime.Loader;
+using System.Runtime.Versioning;
 
 namespace Internal.Runtime.InteropServices
 {
     /// <summary>
     /// This class enables the .NET IJW host to load an in-memory module as a .NET assembly
     /// </summary>
+    [SupportedOSPlatform("windows")]
     public static class InMemoryAssemblyLoader
     {
-#if TARGET_WINDOWS
         private static bool IsSupported { get; } = InitializeIsSupported();
-
         private static bool InitializeIsSupported() => AppContext.TryGetSwitch("System.Runtime.InteropServices.EnableCppCLIHostActivation", out bool isSupported) ? isSupported : true;
-#endif
 
         /// <summary>
-        /// Loads into an isolated AssemblyLoadContext an assembly that has already been loaded into memory by the OS loader as a native module.
+        /// Loads an assembly that has already been loaded into memory by the OS loader as a native module
+        /// into an isolated AssemblyLoadContext.
         /// </summary>
         /// <param name="moduleHandle">The native module handle for the assembly.</param>
         /// <param name="assemblyPath">The path to the assembly (as a pointer to a UTF-16 C string).</param>
+        [RequiresUnreferencedCode("C++/CLI is not trim-compatible", Url = "https://aka.ms/dotnet-illink/nativehost")]
```

Comment:
I'm pretty sure this will generate a warning for an empty console app... since it's rooted always.
The RUC should be left on the `LoadInMemoryAssemblyInContextImpl` only. The warning that produces in dotnet/runtime build should be suppressed by pragmas (for the analyzer) and by the XML suppression for the dotnet/runtime trimming during build (see how this is done before the change).

When trimming an application, the `IsSupported` is set to false and the trimmer will only see the `throw` branch and it will remove the call to `LoadInMemoeryAssemblyInContextImpl`, thus not producing any warning because of it.
So in trimmed app we're left with this method body having just `throw`.

Reply:
Looks like it is 'Elinor does not understand trimming' time again...

I tried to follow what `ComActivator.GetClassFactoryForType` does with `RequiresUnreferencedCode` on the method and the unconditional rooting: https://github.com/dotnet/runtime/blob/84de1282388f5c2506413326f2d49b6abec677df/src/coreclr/System.Private.CoreLib/src/Internal/Runtime/InteropServices/ComActivator.cs#L92-L93
https://github.com/dotnet/runtime/blob/84de1282388f5c2506413326f2d49b6abec677df/src/coreclr/System.Private.CoreLib/src/ILLink/ILLink.Descriptors.Windows.xml#L7-L9
What am I missing that makes the RUC work fine there? Is it from being non-`public`? Or `UnmanagedCallersOnly`? If so, can we just change this to match, since it is only in SPCL (I don't know why it is public in the first place)?

I was expecting that one of the `*.TrimmingTests` would basically be an empty console and fail for warnings. I guess we don't have that and just always set `SuppressTrimAnalysisWarnings=true` with none of the tests overriding it? Seems like it might be a useful basic test to easily catch people like me 😅

Reply:
So in a continuation of not knowing what I'm doing, I tried (with the changes in this PR):
1. Take one of the trimming tests (TypeBuilderComDisabled)
2. Explicitly set `EnableTrimAnalyzer=true` and `SuppressTrimAnalysisWarnings=false`
3. Publish trimmed
4. [no warnings]
5. Set `BuiltInComInteropSupport=true`
6. [no warnings]
7. Set `EnableCppCLIHostActivation=true`
8. [no warnings] 

I don't recall - did we intentionally make COM not warn?

If I set something else like `StartupHookSupport=true`, it does warn, so I think I enabled the warnings correctly. And without my changes, setting `EnableCppCLIHostActivation=true` does warn, so that is something I broke.

Reply:
cc @sbomer @LakshanF

Reply:
> I was expecting that one of the *.TrimmingTests would basically be an empty console and fail for warnings. I guess we don't have that and just always set SuppressTrimAnalysisWarnings=true with none of the tests overriding it? Seems like it might be a useful basic test to easily catch people like me

This seems like a reasonable expectation. Not sure we have that, but if we don't, we should add it.

Reply:
Sorry - I obviously don't know linker well enough - even though I write the code involved in this. In short - methods which are kept via the descriptor XML will NOT produce warnings (intentional behavior in linker right now - I remember the reasons only vaguely - the comment says "Don't want to warn as it's intentional choice"... definitely not good enough description).
So the reason why this works for StartupHook for example is because the entry point (`ProcessStartupHook`) is actually not marked as RUC (and can be rooted unconditionally), but it will call `CallStartupHook` if the feature switch is enabled which is RUC annotated - so that will warn. So we don't need to use features switches in the descriptors, we can include the entry points unconditionally. This has two benefits (which I forgot about)
* If the app does end up calling the entry point - it will get a nice descriptive exception (unlike something like MissingMethodException which is not very helpful)
* It produces trim warning if the feature switch is enabled

If I remember correctly we removed this unconditional rooting for the native host entry point because there you only get an error code back anyway (so the rich exception is lost regardless) and Marek really wanted to remove the potentially unnecessary type/method/exception string.
I guess we could do the same for the rest - they report error codes in reality anyway.

So I would stick to unconditionally rooting the entry point (like before) - not annotate with RUC (it's cleaner, than relying on a weird linker behavior). Annotated the inner method. Suppress the warning via pragma/librarbuildsuppression.xml.

Reply:
I think the lack of warning for COM (and why my previous change removed the worningfor C++/CLI) is actually because we have the function marked with `RequiresUnreferencedCode`? So there's no warning on those functions themselves for using another function marked with RUC and they would never be called explicitly from managed, so users never get a warning even with `BuiltInComInteropSupport=true`.

Reply:
That might be true... in which case it's definitely broken. Yes - this looks broken - publishing an app with `BuiltInComInteropSupport=true` should warn - I think... it doesn't. @LakshanF do you remember the details here?
I would expect to get a warning even if the app itself doesn't actually call any COM APIs - because enabling the feature switch also enables the app to host managed COM components (via the native entry points).

Reply:
> I guess we could do the same for the rest - they report error codes in reality anyway.

Right now (without this PR), for if you activate C++/CLI in a trimmed managed app, you get
`System.NotSupportedException: This API is not enabled in trimmed scenarios. see https://aka.ms/dotnet-illink/nativehost for more details`.

The actual message could be more useful / say something about C++/CLI being disabled, but it is a nicer error than just `SEHException`.

I didn't try COM, but I expect/hope it would have `System.NotSupportedException: Built-in COM has been disabled via a feature switch. See https://aka.ms/dotnet-illink/com for more information.`